src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointMatcherPolicy.cs

```diff
@@ -0,0 +1,113 @@
+            }
+
+            if (!ContainsDynamicEndpoints(endpoints))
+            {
+                // Dynamic page endpoints are always dynamic endpoints.
+                return false;
+            }
+
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var metadata = endpoints[i].Metadata.GetMetadata<DynamicPageMetadata>();
+                if (metadata != null)
+                {
+                    // Found a dynamic page endpoint
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public async Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            // There's no real benefit here from trying to avoid the async state machine.
+            // We only execute on nodes that contain a dynamic policy, and thus always have
+            // to await something.
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                var endpoint = candidates[i].Endpoint;
+
+                var metadata = endpoint.Metadata.GetMetadata<DynamicPageMetadata>();
+                if (metadata == null)
+                {
+                    continue;
+                }
+
+                var matchedValues = candidates[i].Values;
+                var endpoints = await _selector.SelectEndpointsAsync(metadata.Values);
+                if (endpoints.Count == 0)
+                {
+                    // If there's no match this is a configuration error. We can't really check
+                    // during startup that the action you configured exists.
+                    throw new InvalidOperationException(
+                        "Cannot find the fallback endpoint specified by route values: " + 
+                        "{ " + string.Join(", ", metadata.Values.Select(kvp => $"{kvp.Key}: {kvp.Value}")) + " }.");
+                }
+
+                var replacement = endpoints[0];
+                
+                // We need to provide the route values associated with this endpoint, so that features
+                // like URL generation work.
+                var values = new RouteValueDictionary(metadata.Values);
+
+                // Include values that were matched by the fallback route.
+                foreach (var kvp in matchedValues)
+                {
+                    values.TryAdd(kvp.Key, kvp.Value);
+                }
+
+                candidates.ReplaceEndpoint(i, replacement, values);
+            }
+        }
+    }
+}
```

Comment:
There's a certain amount of de-duplication that could be done between these two pages (controller and pages). 

I'm not incredibly keen because to a certain extent these code paths need to each have their own types for everything because of DI/metadata. Open to suggestions/comments 😆

Reply:
Make it generic based on the selector and the metadata type :trollface:

Reply:
Resolving this comment. I feel like I solved this for the most part by factoring more code into the action selection table class.